src/Specular/Internal/Incremental.purs

```diff
@@ -0,0 +1,410 @@
+removeObserver = mkEffectFn2 \node observer -> do
+  oldRefcount <- runEffectFn1 Node.refcount node
+  observers <- runEffectFn1 Node.get_observers node
+  runEffectFn2 MutableArray.remove observers observer
+  runEffectFn2 handleRefcountChange node oldRefcount
+
+addDependent :: forall a. EffectFn2 (Node a) SomeNode Unit
+addDependent = mkEffectFn2 \node dependent -> do
+--  trace $ "addDependent " <> show (Node.name' node) <> " -> " <> show (Node.name' dependent)
+
+  oldRefcount <- runEffectFn1 Node.refcount node
+  dependents <- runEffectFn1 Node.get_dependents node
+  runEffectFn2 MutableArray.push dependents dependent
+  runEffectFn2 handleRefcountChange node oldRefcount
+
+removeDependent :: forall a. EffectFn2 (Node a) SomeNode Unit
+removeDependent = mkEffectFn2 \node dependent -> do
+  oldRefcount <- runEffectFn1 Node.refcount node
+  dependents <- runEffectFn1 Node.get_dependents node
+  runEffectFn2 MutableArray.remove dependents dependent
+  runEffectFn2 handleRefcountChange node oldRefcount
+
+handleRefcountChange :: forall a. EffectFn2 (Node a) Int Unit
+handleRefcountChange = mkEffectFn2 \node oldRefcount -> do
+  newRefcount <- runEffectFn1 Node.refcount node
+  if oldRefcount == 0 && newRefcount > 0 then
+    runEffectFn1 connect node
+  else if oldRefcount > 0 && newRefcount == 0 then
+    runEffectFn1 disconnect node
+  else
+    pure unit
+
+  -- Update globalTotalRefcount
+  oldTotalRefcount <- runEffectFn1 Ref.read globalTotalRefcount
+  runEffectFn2 Ref.write globalTotalRefcount (oldTotalRefcount - oldRefcount + newRefcount)
+
+-- Preconditions:
+-- - node does not have value computed
+-- - node does not have any dependents
+--
+-- Postconditions:
+-- - all dependencies are connected and have value computed
+-- - node has value computed
+-- - node has correct height
+connect :: forall a. EffectFn1 (Node a) Unit
+connect = mkEffectFn1 \node -> do
+--  trace $ "connect " <> show (Node.name' node)
+
+  source <- runEffectFn1 Node.get_source node
+  dependencies <- source.dependencies
+
+  runEffectFn2 Array.iterate dependencies $ mkEffectFn1 \dependency -> do
+    runEffectFn2 addDependent dependency (toSomeNode node)
+    dependencyHeight <- runEffectFn1 Node.get_height dependency
+    ourHeight <- runEffectFn1 Node.get_height node
+    if dependencyHeight + 1 > ourHeight then do
+      runEffectFn2 Node.set_height node (dependencyHeight + 1)
+      runEffectFn2 Node.set_adjustedHeight node (dependencyHeight + 1)
+    else
+      pure unit
+
+--  trace $ "connect: node " <> show (Node.name' node) <> ": compute"
+  value <- runEffectFn1 source.compute node
+  runEffectFn2 Node.set_value node value
+
+disconnect :: forall a. EffectFn1 (Node a) Unit
+disconnect = mkEffectFn1 \node -> do
+  source <- runEffectFn1 Node.get_source node
+
+  dependencies <- source.dependencies
+  runEffectFn2 Array.iterate dependencies $ mkEffectFn1 \dependency -> do
+    runEffectFn2 removeDependent dependency (toSomeNode node)
```

Comment:
Sorry if this was answered in presentation: when disconnecting a node, shouldn't we adjust it's heights?

Reply:
See https://github.com/restaumatic/purescript-specular/pull/66/files#diff-6fcd89c10cb29e5a730a6b6ea2db4615efbcf6d65d177d8441f8623d82742e85R45

> Node height is maintained only when the node is connected.

So we could, but we don't have to.

Besides, what would we change the height to?